Builder/TreeBuilder.php

```diff
@@ -0,0 +1,246 @@
+     *
+     * @return Node[] the array is indexed by route name
+     */
+    public function buildTree ($routeCollection)
+    {
+        $relevantRoutes = $this->calculateRelevantRoutes($routeCollection);
+        $nodes = $this->generateNodesFromRoutes($routeCollection, $relevantRoutes);
+        $nodes = $this->linkNodeHierarchy($nodes);
+
+        // needs to be after the hierarchy has been set up
+        return $this->calculateAllParameters($nodes);
+    }
+
+
+
+    /**
+     * Calculates a list of which routes should be included in the tree
+     *
+     * @param Route[]|RouteCollection $routeCollection
+     *
+     * @return array of format ["route" => (bool) $includeInTree]
+     * @throws InvalidRouteTreeException
+     */
+    private function calculateRelevantRoutes ($routeCollection)
+    {
+        $routeIndex = [];
+
+        foreach ($routeCollection as $routeName => $route)
+        {
+            $routeIndex[$routeName] = false;
+        }
+
+        foreach ($routeCollection as $routeName => $route)
+        {
+            $routeData = $route->getOption(self::ROUTE_OPTIONS_KEY);
+
+            // no route data found -> skip
+            if (null === $routeData)
+            {
+                continue;
+            }
+
+            // route data found -> mark as relevant
+            $routeIndex[$routeName] = true;
+
+            // mark parent route as relevant
+            if (isset($routeData["parent"]) && !empty($routeData["parent"]))
+            {
+                $parentRoute = $routeData["parent"];
+
+                if (!isset($routeIndex[$parentRoute]))
+                {
+                    throw new InvalidRouteTreeException(sprintf(
+                        "Route '%s' references a parent '%s', but the parent route does not exist.",
+                        $routeName,
+                        $parentRoute
+                    ));
+                }
+
+                $routeIndex[$parentRoute] = true;
+            }
+        }
+
+        return $routeIndex;
+    }
+
+
+
+    /**
+     * Generates the nodes for the given routes
+     *
+     * @param Route[]|RouteCollection $routeCollection
+     * @param array                   $relevantRoutes
+     *
+     * @return Node[]
+     */
+    private function generateNodesFromRoutes ($routeCollection, array $relevantRoutes)
+    {
+        $nodes = [];
+
+        // generate nodes fro
+        foreach ($routeCollection as $routeName => $route)
+        {
+            // skip not relevant routes
+            if (!isset($relevantRoutes[$routeName]) || !$relevantRoutes[$routeName])
+            {
+                continue;
+            }
+
+            $nodes[$routeName] = $this->createNodeFromRoute($routeName, $route);
+        }
+
+        return $nodes;
+    }
+
+
+
+    /**
+     * Links the hierarchy between the nodes
+     *
+     * @param Node[] $nodes
+     *
+     * @return Node[]
+     */
+    private function linkNodeHierarchy ($nodes)
```

Comment:
Yes. I removed all the typehints from `$routes` in this class, but apparently removed this one too.